python/lsst/daf/butler/core/dimensions/graph.py

```diff
@@ -0,0 +1,473 @@
+
+    def __getitem__(self, key):
+        return self._dimensions[key]
+
+    def get(self, key, default=None):
+        """Return the element with the given name, or ``default`` if it
+        does not exist.
+
+        ``key`` may also be a `Dimension`, in which case an equivalent
+        object will be returned.
+        """
+        return self._dimensions.get(key, default)
+
+    @property
+    def names(self):
+        """The names of all elements (`set`-like, immutable).
+
+        The order of the names is consistent with the iteration order of the
+        `DimensionSet` itself.
+        """
+        return self._dimensions.names
+
+    @property
+    def links(self):
+        """The names of all fields that uniquely identify these dimensions in
+        a data ID dict (`frozenset` of `str`).
+        """
+        return self._dimensions.links
+
+    def __hash__(self):
+        return hash(self._dimensions)
+
+    def __eq__(self, other):
+        return self._dimensions == other
+
+    def __le__(self, other):
+        return self._dimensions <= other
+
+    def __lt__(self, other):
+        return self._dimensions < other
+
+    def __ge__(self, other):
+        return self._dimensions >= other
+
+    def __gt__(self, other):
+        return self._dimensions > other
+
+    def issubset(self, other):
+        """Return `True` if all dimensions in ``self`` are also in ``other``.
+
+        The empty graph is a subset of all graphs (including the empty graph).
+        """
+        return self <= other
+
+    def issuperset(self, other):
+        """Return `True` if all dimensions in ``other`` are also in ``self``,
+        and `False` otherwise.
+
+        All graphs (including the empty graph) are supersets of the empty
+        graph.
+        """
+        return self >= other
+
+    def isdisjoint(self, other):
+        """Return `True` if there are no dimensions in both ``self`` and
+        ``other``, and `False` otherwise.
+
+        All graphs (including the empty graph) are disjoint with the empty
+        graph.
+        """
+        return self._dimensions.isdisjoint(other)
+
+    def union(self, *others, optional=False):
+        """Return a new graph containing all elements that are in ``self`` or
+        any of the other given graphs.
+
+        Parameters
+        ----------
+        *others : iterable over `Dimension` or `str`.
+            Other sets whose elements should be included in the result.
+
+        Returns
+        -------
+        result : `DimensionGraph`
+            Graph containing any elements in any of the inputs.
+        """
+        return DimensionGraph(self.universe, self._dimensions.union(*others), optional=optional)
+
+    def intersection(self, *others, optional=False):
+        """Return a new graph containing all elements that are in both  ``self``
+        and all of the other given graphs.
+
+        Parameters
+        ----------
+       *others : iterable over `Dimension` or `str`.
```

Comment:
```rst
*others : `Dimension` or `str`
    Other sets whose elements may be included in the result.
```

Remember that `others` is an iterable only internally. To a user, they pass `Dimension` or `str` instances as positional arguments.